### PR TITLE
[libc++] Implement std::gcd using the binary version

### LINKED ISSUE #77648: [D145982] [libc++] Implement std::gcd using the binary version
This issue tracks picking up https://reviews.llvm.org/D145982 from the Phabricator archive.

### PR DESCRIPTION
**Pull request by serge-sans-paille**
The binary version is four times faster than current implementation in my setup, and generally considered a better implementation.

Code inspired by https://en.algorithmica.org/hpc/algorithms/gcd/ which itself is inspired by https://lemire.me/blog/2013/12/26/fastest-way-to-compute-the-greatest-common-divisor/

Fix #77648

### CODE REVIEW

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.gcd/gcd.pass.cpp
Comment (AdvenamTacet) [inline libcxx/test/std/numerics/numeric.ops/numeric.ops.gcd/gcd.pass.cpp:168]: Because [`std::gcd` works for negative numbers](https://godbolt.org/z/dfEbdEWzW), we should test also signed types.

**Review by AdvenamTacet — CHANGES_REQUESTED** [review]
Looks good to me % testing.
To be sure that we don't change `std::gcd` behavior, let's test (compare with naive gcd) also:
- minimal and maximal values for different types.
- all pairs of numbers around 0.
```cpp
    for(T i = -9; i < 10; ++i)
        for(T j = -9; j < 10; ++j)
```

I don't see a benchmark [requested on Phabricator](https://reviews.llvm.org/D145982#inline-1519289).

@philnik777 do changes in `transitive_includes` look good to you?

File: libcxx/include/__numeric/gcd_lcm.h
Comment (AdvenamTacet) [inline libcxx/include/__numeric/gcd_lcm.h:44]: Please, add a comment explaining why it's done that way.

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.gcd/gcd.pass.cpp
Comment (AdvenamTacet) [inline libcxx/test/std/numerics/numeric.ops/numeric.ops.gcd/gcd.pass.cpp:63]: Old `gcd` implementation had no ifs, from what I can see. I think, we should test if the new implementation is 1:1 with the previous one. Or if we change behavior, we should state it explicitly in PR with reasoning. In that case, a comment with explanation would be good here.

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.gcd/gcd.pass.cpp
Comment (AdvenamTacet) [inline libcxx/test/std/numerics/numeric.ops/numeric.ops.gcd/gcd.pass.cpp:84]:
```suggestion
      std::numeric_limits<Input>::min(),
      std::numeric_limits<Input>::min() + 1,
      std::numeric_limits<Input>::min() + 2,
      std::numeric_limits<Input>::max(),
      std::numeric_limits<Input>::max() - 1,
      std::numeric_limits<Input>::max() - 2,
```

File: libcxx/include/__numeric/gcd_lcm.h
Comment (AdvenamTacet) [inline libcxx/include/__numeric/gcd_lcm.h:73]: Could you add a comment explaining the algorithm and with links to further materials?

**Review by AdvenamTacet — CHANGES_REQUESTED** [review]
`gcd` implementation and `gcd` test look good to me with addressed comments.

Please, add a benchmark in `libcxx/benchmarks/`.

**Comment by AdvenamTacet** [comment]
Also, while for random values performance gain is consistent, it's worth to point out that it's possible to create benchmarks in which naive GCD is clearly faster. https://quick-bench.com/q/1y8T00mwK3t9EL__IqTKLgmTHwI

We may want to consider running naive GCD if one value is very small.

**Comment by AdvenamTacet** [comment]
Until now, I was not really thinking how this algorithm works, but I finally looked at it closer and I don't think we should upstream it.

The idea is based on:
```
If power of two divides both numbers, we can push it out.
- gcd( 2^x * a, 2^x * b) = 2^x * gcd(a, b)

If and only if exactly one number is even, we can divide that number by  that power. 
- if a, b are odd, then gcd(2^x * a, b) = gcd(a, b)

And standard gcd algorithm where instead of modulo, minus is used.
```

Based on that knowledge, it's possible to create an example where the new GCD is hundred times slower. https://quick-bench.com/q/4_kHmlvAN7nwcDaMqk3_6BSGaRM

Moreover, we still can build a test like that, even if we exclude very small values. This algorithm works well for random values, but unfortunately not for all values. https://quick-bench.com/q/vUh_PAKAip6GLvnH3FvgnmzdK74

Potential performance hit is so big that I don't think we can upstream it.

**Comment by philnik777** [comment]
@AdvenamTacet You always have to be careful when writing benchmarks. If you look at the generated code, you see that Clang was able to constant-fold the native version and you only benchmark moving `0x1` into memory. While we should also consider this, it's not exactly a common case. If you avoid that the numbers look much better: https://quick-bench.com/q/KxLGz_xNDsVCEd3UUgBD6R9oPlY.
Much more importantly: I can produce similarly bad cases for just about any optimziation in the library. The question is how likely it is that this is an actual problem in the wild. I'd guess that it's not very common to calculate the gcd of `(2^n)-1` and 1.
Given that, there are a few questions to answer:
- In which cases is it faster or slower?
- How often do we expect these cases to occur?
- Would it be possible to detect these cases and decide which algorithm to use based on that? (e.g. with a `popcount`)
- How much of a difference does it make if we don't have perfect branch prediction?

**Comment by AdvenamTacet** [comment]
@philnik777 thx for looking at it and fixing my benchmark! It's still big performance hit, but you are right that it's probably unlikely to see it in the wild often.

> - In which cases is it faster or slower?
> - Would it be possible to detect these cases and decide which algorithm to use based on that? (e.g. with a popcount)

My fast tests show that for random numbers or interval of numbers, it's unlikely to have worse performance with this new implementation when  `abs(__a) > 250 && abs(__b) > 250`. Probably we can do something similar with a number chosen more carefully.

> How often do we expect these cases to occur?

On every big interval I tested, implementation from this PR is faster, I did not try to check how often it's slower and how often it's faster. I can try to look at it soon.
Probably we should look at discrete logarithm for a better estimation.

On average, it may have a positive effect, if we are ok with big slowdown in some rare cases.
https://grep.app/search?q=std%3A%3Agcd%28
https://grep.app/search?current=3&q=gcd%28

**Comment by lemire** [comment]
> it's possible to create an example where the new GCD is hundred times slower

The algorithmic complexity of the binary GCD is the number of bits in the largest of the two integers

For the benchmark in question, it should go down as...

```
u: 1, v: 16777214
u: 1, v: 8388606
u: 1, v: 4194302
u: 1, v: 2097150
u: 1, v: 1048574
u: 1, v: 524286
u: 1, v: 262142
u: 1, v: 131070
u: 1, v: 65534
u: 1, v: 32766
u: 1, v: 16382
u: 1, v: 8190
u: 1, v: 4094
u: 1, v: 2046
u: 1, v: 1022
u: 1, v: 510
u: 1, v: 254
u: 1, v: 126
u: 1, v: 62
u: 1, v: 30
u: 1, v: 14
u: 1, v: 6
u: 1, v: 2
```

So it is difficult to imagine that that there could be gigantic differences.

**Comment by lemire** [comment]
Note that my statement is not meant to contradict @AdvenamTacet but to point out that you can actually bound the difference from first principles. Algorithmically speaking, the difference cannot be excessively large, especially because each step in the binary GCD is cheaper. So it is definitively true that the naive algorithm can be sometimes faster, but there is denial-of-service attack lurking. :-)

**Comment by ylchapuy** [comment]
An hybrid approach with one modulo at the begining followed by the binary gcd loop could get us the best of both world I think.

Here a proposed implementation:
```
template <class T>
T binary_gcd(T u, T v) {
  if (u < v) std::swap(u, v);
  if (v == 0)
    return u;
  u %= v; // make both argument of the same size, and early result in the easy case
  if (u == 0)
    return v;

  auto zu = std::countr_zero(u);
  auto zv = std::countr_zero(v);
  auto shift = std::min(zu, zv);
  u >>= zu;
  v >>= zv;
  do {
    T u_minus_v = u - v;
    if (u > v)
      u = v, v = u_minus_v;
    else
      v = v - u;
    // v >>= std::countr_zero(v); !!! avoid instruction dependency
    v >>= std::countr_zero(u_minus_v);
  } while(v != 0);
  return u << shift;
}
```
And here some benchmarks: https://quick-bench.com/q/AMjI0bBfq3YCyz6XBtvINUNQsBs

File: libcxx/include/__numeric/gcd_lcm.h
Comment (AdvenamTacet) [reply libcxx/include/__numeric/gcd_lcm.h:73]: I would be happy, if that information is included:
```
If power of two divides both numbers, we can push it out.
- gcd( 2^x * a, 2^x * b) = 2^x * gcd(a, b)

If and only if exactly one number is even, we can divide that number by  that power. 
- if a, b are odd, then gcd(2^x * a, b) = gcd(a, b)

And standard gcd algorithm where instead of modulo, minus is used.
```

**Comment by AdvenamTacet** [comment]
That change looks good to me and resolves my performance concerns. Overall, LGTM with my previous in-code comments addressed.

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.gcd/gcd.pass.cpp
Comment (AdvenamTacet) [reply libcxx/test/std/numerics/numeric.ops/numeric.ops.gcd/gcd.pass.cpp:63]: I see it's done.

File: libcxx/include/__numeric/gcd_lcm.h
Comment (serge-sans-paille) [reply libcxx/include/__numeric/gcd_lcm.h]: Done!

**Comment by serge-sans-paille** [comment]
@AdvenamTacet does this version look good to you?

**Review by AdvenamTacet — APPROVED** [review]
I have no more performance concerns and LGTM % two hanging comments.

Please add to test values numbers around limits instead of only limits:
```cpp
      std::numeric_limits<Input>::min(),
      std::numeric_limits<Input>::min() + 1,
      std::numeric_limits<Input>::min() + 2,
      std::numeric_limits<Input>::max(),
      std::numeric_limits<Input>::max() - 1,
      std::numeric_limits<Input>::max() - 2,
```
could you also comment on that change:
```diff
-      return -static_cast<_Result>(__t);
+      return static_cast<_Result>(-static_cast<std::make_unsigned_t<_Result>>(__t));
```
Why that change? When does it change the result?

**Comment by serge-sans-paille** [comment]
> I have no more performance concerns and LGTM % two hanging comments.
> 
> Please add to test values numbers around limits instead of only limits:
> 
> ```c++
>       std::numeric_limits<Input>::min(),
>       std::numeric_limits<Input>::min() + 1,
>       std::numeric_limits<Input>::min() + 2,
>       std::numeric_limits<Input>::max(),
>       std::numeric_limits<Input>::max() - 1,
>       std::numeric_limits<Input>::max() - 2,
> ```

done

> could you also comment on that change:
> 
> ```diff
> -      return -static_cast<_Result>(__t);
> +      return static_cast<_Result>(-static_cast<std::make_unsigned_t<_Result>>(__t));
> ```
> 
> Why that change? When does it change the result?

I wasn't aware that gcd is UB when one of its argument absolute value cannot be represented in the result type. I dismissed this change.

**Comment by serge-sans-paille** [comment]
@AdvenamTacet I'm going to merge this tomorrow, if that's fine with you?

**Comment by hiraditya** [comment]
We should test this algorithm with ubsan

**Pull request merged**
